### DISCUSSION TITLE
Using pvwatts losses with ModelChain

### DISCUSSION THREAD

**Discussion by casaout** [pvlib/pvlib-python#1357] [Q&A] [answered]
Hi

I am trying to use the [`pvwatts loss model`](https://pvlib-python.readthedocs.io/en/stable/generated/pvlib.pvsystem.pvwatts_losses.html) with the [`ModelChain`](https://pvlib-python.readthedocs.io/en/v0.9.0/generated/pvlib.modelchain.ModelChain.html), similar to [Jawairia123 a few years ago](https://github.com/pvlib/pvlib-python/issues/484). In the meantime, there was a [successful PR](https://github.com/pvlib/pvlib-python/pull/491) by @cwhanse and @wholmgren, to make this possible.

Unfortunately, I am unsuccessful at correctly applying the helper-function at this point and haven't been able to find a working example online.

**Question 1**: Input parameter units? Per [documentation](https://pvlib-python.readthedocs.io/en/stable/generated/pvlib.pvsystem.pvwatts_losses.html), the input and output unit is in`%`. However, the default parameters range from 0-3 (which would be 0-300%). How can I make sense of this?
Concretely, if a pvsystem is e.g. 8years old, relative to what do I consider it to receive an input  `%`? 
Also, to understand the other parameters, what would `shading==3` or `availability==3` (all the defaults) mean? 

**Question 2**:  How do I add the `losses_model` to the ModelChain? What I am trying today:

    system_losses = pvwatts_losses(age=8, snow=0)  # todo: % or number of years? (defaults seem absolute, not %)
    print(f"LOSS: {1 - (system_losses / 100)}")
    mc = ModelChain(system, location, aoi_model="physical", losses_model=system_losses) 
    mc.run_model(times=mc_weather.index, weather=mc_weather)

However, the output is: that the `first argument must be callable`. An alternative would be what [Jawairia123 suggested](https://github.com/pvlib/pvlib-python/issues/484#issuecomment-398385156), but my assumption was that the PR mentioned above made it possible in the meantime to use the pvwatts loss model with ModelChain.

Any pointers are appreciated. thank you!

**Comment by kandersolar** [comment, marked as answer]
Hi @casaout, hopefully this clears things up a bit :)

> Question 1: Input parameter units? Per documentation, the input and output unit is in %. However, the default parameters range from 0-3 (which would be 0-300%). How can I make sense of this?

The inputs are in `%`, meaning a default value of 3 is interpreted as 3%=0.03.  If the inputs were supposed to be unitless then the default value would be `shading=0.03`. 

---

> Also, to understand the other parameters, what would shading==3 or availability==3 (all the defaults) mean?

Each of these are percentage losses that get combined into a single overall loss fraction.  So for example `shading=3` means that 3% of system production is lost to shading, or the system produces only 97% of what it would have if `shading=0`.  And the other losses stack in a similar fashion.  Note that PVWatts losses are applied to the overall system output, not at the point in the system that each loss occurs in real life -- so for example, a soiling loss is not "masked" by inverter clipping in the PVWatts loss model. 

---

> system_losses = pvwatts_losses(age=8, snow=0)  # todo: % or number of years? (defaults seem absolute, not %)

The `age` parameter is not a number of years, but simply another percentage loss like all the other parameters.  So for example, if you expect the system's performance to degrade at 0.5%/yr, and the system is 8 years old, then you might specify `age=4` to indicate a 4% loss due to system age.  Anything more sophisticated than that (e.g. an `age` loss that changes with time) is outside the scope of a simple model like PVWatts.

---

> print(f"LOSS: {1 - (system_losses / 100)}")

Note that `pvwatts_losses` returns a loss (not the remaining capacity), so e.g. with the default inputs, the return value of ~14 represents a 14% loss, or an 86% remaining capacity.  So this `print` line of code is not correct, as it converts the loss to a remaining capacity but then calls it a loss. 

---

> mc = ModelChain(system, location, aoi_model="physical", losses_model=system_losses

Check out the `ModelChain` [docs](https://pvlib-python.readthedocs.io/en/stable/generated/pvlib.modelchain.ModelChain.html) to see what is expected for the `losses_model` parameter:

> losses_model (str or function, default 'no_loss') – Valid strings are ‘pvwatts’, ‘no_loss’. The ModelChain instance will be passed as the first argument to a user-defined function.

So the `losses_model` value is supposed to be a string (or a function if you want to customize it, but nevermind that) like how you have `aoi_model="physical"`.  To specify the loss percentages you would use the `losses_parameters` input to `PVSystem` rather than calling the `pvwatts_losses` function yourself.  In fact this is how you specify parameters for any model: set them in the appropriate `PVSystem` dict, then let `ModelChain` call the functions for you.

Here is a short example for using the `losses_parameters` input to `PVSystem` -- change the value for `age`, or specify any other pvwatts loss in that dict, and the `ac` output of the `ModelChain` will change accordingly. 

```python
from pvlib.modelchain import ModelChain
from pvlib.pvsystem import PVSystem
from pvlib.location import Location
import pandas as pd

system = PVSystem(losses_parameters={'age': 4},
                  module_parameters={'pdc0': 1, 'gamma_pdc': -0.004},
                  inverter_parameters={'pdc0': 1},
                  temperature_model_parameters={'a': -3.56, 'b': -0.075, 'deltaT': 3})
location = Location(latitude=32.2, longitude=-110.9)
weather = pd.DataFrame([[1050, 1000, 100, 30, 5]],
                       columns=['ghi', 'dni', 'dhi', 'temp_air', 'wind_speed'],
                       index=[pd.Timestamp('20170401 1200', tz='US/Arizona')])
mc = ModelChain(system, location, losses_model='pvwatts',
                aoi_model='no_loss', spectral_model='no_loss')
mc.run_model(weather)
print(mc.results.ac)
```

**Reply by casaout** [reply]
Thank you @kanderso-nrel. This definitely clarified things. 
(after being so used of using decimals I've literally not considered that `%` is really just the percentage 😒😁)

Next stop, looking up on the manufacturer website / factsheet how much degradation I can expect per year 😊.
